kolibri/core/content/utils/annotation.py

```diff
@@ -134,6 +135,93 @@ def fix_multiple_trees_with_id_one():
             )
 
 
+def update_num_coach_contents():
+    """
+    Function to set num_coach_content on topic trees that were imported before
+    the annotation would have been run.
+    """
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
+
+    ContentNodeClass = bridge.get_class(ContentNode)
+
+    ContentNodeTable = bridge.get_table(ContentNode)
+
+    connection = bridge.get_connection()
+
+    child = ContentNodeTable.alias()
+
+    logger.info("Updating num_coach_content on existing channels")
+
+    # start a transaction
+
+    trans = connection.begin()
+
+    for channel_id in ChannelMetadata.objects.all().values_list("id", flat=True):
+
+        node_depth = (
+            bridge.session.query(func.max(ContentNodeClass.level))
+            .filter_by(channel_id=channel_id)
+            .scalar()
+        )
+
+        # Update all leaf ContentNodes to have num_coach_content to 1 or 0
+        connection.execute(
+            ContentNodeTable.update()
+            .where(
+                and_(
+                    # In this channel
+                    ContentNodeTable.c.channel_id == channel_id,
+                    # That are not topics
+                    ContentNodeTable.c.kind != content_kinds.TOPIC,
+                )
+            )
+            .values(
+                num_coach_contents=cast(ContentNodeTable.c.coach_content, Integer())
+            )
+        )
```

Comment:
If this is being run for all channels anyway, couldn't this update block be moved outside of the `for` loop and made into a single DB update?

```python
# Update all leaf ContentNodes to have num_coach_content to 1 or 0
connection.execute(
    ContentNodeTable.update()
    .where(ContentNodeTable.c.kind != content_kinds.TOPIC)
    .values(num_coach_contents=cast(ContentNodeTable.c.coach_content, Integer()))
)
```

Reply:
Yes - I just copied it directly from the normal logic and didn't optimize.